spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java

```diff
@@ -385,49 +390,135 @@
-    rewriteTableLikeParameterization(stats, result -> {
+  private void rewriteBinaryTableLikeParameterization(ListIterator<Statement> stats) {
+    int[] i = { 0 };
+    while (stats.hasNext()) {
       Statement stat = stats.next();
       BinaryExpression orExpr = getOrExpression(stat);
       if (orExpr == null) {
         stats.previous();
-        return true;
+        break;
       }
-      splitRow(orExpr, result);
-      return false;
-    });
+      List<Expression> row = new ArrayList<>();
+      splitRow(orExpr, row);
+
+      // replace `a | b || c` by `a ; b ; c` so that they can be handled uniformly later
+      stats.remove();
+      for (int j = 0, rowSize = row.size(); j < rowSize; j++) {
+        Expression expr = row.get(j);
+        Statement exStat = stmt(expr);
+        exStat.setLineNumber(expr.getLineNumber());
+        exStat.setLastLineNumber(expr.getLastLineNumber());
+        exStat.setColumnNumber(expr.getColumnNumber());
+        exStat.setLastColumnNumber(expr.getLastColumnNumber());
+        if ((j == 0) && (stat.getStatementLabels() != null)) {
+          stat.getStatementLabels().forEach(exStat::addStatementLabel);
+        }
+        stats.add(exStat);
+        i[0]++;
+      }
+    }
+
+    if (stats.hasNext()) {
+      boolean nextStatementIsCombineLabeled = COMBINE.equals(stats.next().getStatementLabel());
+      stats.previous();
+      if (!nextStatementIsCombineLabeled) {
+        stats.add(stmt(varX(new DynamicVariable("__", false))));
+        i[0]++;
+      }
+    }
+
+    // rewind to before the added statements
+    for (int j = i[0]; j > 0; j--) {
+      stats.previous();
+    }
   }
 
   private void rewriteExpressionTableLikeParameterization(ListIterator<Statement> stats) throws InvalidSpecCompileException {
-    rewriteTableLikeParameterization(stats, result -> {
+    rewriteTableLikeParameterization(stats, () -> {
+      boolean headerRow;
+      if (stats.hasNext()) {
+        headerRow = COMBINE.equals(stats.next().getStatementLabel());
+        stats.previous();
+      } else {
+        headerRow = false;
+      }
+
+      if (headerRow) {
+        rewriteBinaryTableLikeParameterization(stats);
+      }
+
       List<Expression> row = getExpressionChain(stats);
       if (row.size() <= 1) {
         // rewind
         row.forEach(expression -> stats.previous());
-        return true;
+        return null;
       }
-      result.addAll(row);
-      return false;
+      return new DataTableRow(row, headerRow);
     });
   }
 
-  private void rewriteTableLikeParameterization(ListIterator<Statement> stats, Function<List<Expression>, Boolean> rowExtractor) throws InvalidSpecCompileException {
+  private void rewriteTableLikeParameterization(
+    ListIterator<Statement> stats,
+    ThrowingSupplier<DataTableRow, InvalidSpecCompileException> rowExtractor) throws InvalidSpecCompileException {
```

Comment:
Why do you use a supplier instead of a `ThrowingFunction<ListIterator<Statement>, DataTableRow, InvalidSpecCompileException>` I think this would make the for a cleaner relationship.